### PR TITLE
Use common dispatcher for public client outbound

### PR DESCRIPTION
**Pull request by vytautas-karpavicius**
**What changed?**
Stop using `DispatcherProvider` for creating public client. Instead use common dispatcher from rpc.Factory.

Builder for public client outbound was created which also includes auth middleware initialization. It was previously located within server.go itself. Moving it here allows consolidating all logic related logic in one place and covering it with unit tests.

**Why?**
To eventually delete `DispatcherProvider` entirely.

**How did you test it?**
Existing tests and new unit tests.

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: common/rpc/params_test.go [hunk 1]
```diff
@@ -45,6 +47,9 @@ func TestNewParams(t *testing.T) {
 	_, err = NewParams(serviceName, makeConfig(config.Service{RPC: config.RPC{BindOnIP: "invalidIP"}}))
 	assert.EqualError(t, err, "failed to get listen IP: unable to parse bindOnIP value or it is not an IPv4 address: invalidIP")
 
+	_, err = NewParams(serviceName, &config.Config{Services: map[string]config.Service{"frontend": {}}})
+	assert.EqualError(t, err, "failed to create public client outbound: need to provide an endpoint config for PublicClient")
```
Comment (mantas-sidlauskas) [inline common/rpc/params_test.go:51]: nit: error messages can be shorter. Ref: https://github.com/uber-go/guide/blob/master/style.md#error-wrapping

File: common/rpc/outbounds.go [hunk 2]
```diff
@@ -21,12 +21,57 @@
...
+
+func newPublicClientOutbound(config *config.Config) (publicClientOutbound, error) {
+	if len(config.PublicClient.HostPort) == 0 {
+		return publicClientOutbound{}, fmt.Errorf("need to provide an endpoint config for PublicClient")
+	}
+
+	var authMiddleware *authOutboundMiddleware
+	if config.Authorization.OAuthAuthorizer.Enable {
+		clusterName := config.ClusterGroupMetadata.CurrentClusterName
+		clusterInfo := config.ClusterGroupMetadata.ClusterGroup[clusterName]
```
Comment (mantas-sidlauskas) [inline common/rpc/outbounds.go:54]: do we need to check if `clusterName` is in map?

File: common/rpc/outbounds.go [hunk 3]
```diff
@@ -21,12 +21,57 @@
...
 	Build(*grpc.Transport, *tchannel.Transport) (yarpc.Outbounds, error)
 }
+
+func newPublicClientOutbound(config *config.Config) (publicClientOutbound, error) {
+	if len(config.PublicClient.HostPort) == 0 {
+		return publicClientOutbound{}, fmt.Errorf("need to provide an endpoint config for PublicClient")
+	}
+
+	var authMiddleware *authOutboundMiddleware
+	if config.Authorization.OAuthAuthorizer.Enable {
```
Comment (mantas-sidlauskas) [inline common/rpc/outbounds.go:52]: nit/optional: reduce nesting with early return when `Enabled == false`

**Review by mantas-sidlauskas — APPROVED** [review]
lgtm, All suggestions are optional

File: common/rpc/outbounds.go (hunk 2 above)
Comment (vytautas-karpavicius) [reply common/rpc/outbounds.go:54]: This is validated here: https://github.com/uber/cadence/blob/master/common/config/cluster.go#L111
So we should be ok.

File: common/rpc/params_test.go (hunk 1 above)
Comment (vytautas-karpavicius) [reply common/rpc/params_test.go:51]: Agree, will update.

File: common/rpc/outbounds.go (hunk 3 above)
Comment (vytautas-karpavicius) [reply common/rpc/outbounds.go:52]: For this particular case, I would like to keep it. As this is adding optional feature rather than error'ing out early.

**Pull request merged**